[Clang] Ignore invalid base classes

shashforge opened the pull request:
When building a RecordDecl, ignore invalid base classes. This solves crashes later on and leads to better recovery.

Fixes https://github.com/llvm/llvm-project/issues/147186

shashforge committed:
Sema: filter out invalid base-specifiers before attaching
ActOnBaseSpecifiers now skips nullptr entries returned for invalid bases,
avoiding ICE on invalid inheritance. Added regression test in SemaCXX/invalid-inheritance.cpp.

Fixes #147186

Signed-off-by: Shashi Shankar <shashishankar1687@gmail.com>

cor3ntin reviewed (commented):
Thanks for working on this

Did you try  to use `!isUsable` instead of IsInvalid in
`ParseBaseSpecifier` ? I think that might be a simpler fix.

---

This change needs a release note.
Please add an entry to `clang/docs/ReleaseNotes.rst` in the section the most adapted to the change, and referencing any Github issue this change fixes. Thanks!

shashforge committed:
Merge branch 'main' into fix-invalid-base-ice

shashforge committed:
Merge branch 'main' into fix-invalid-base-ice

shashforge committed:
ParseDeclCXX: reject unusable base classes via isUsableType

In Parser::ParseBaseSpecifier, replace the previous `isInvalid` check
with `!BaseTy->isUsableType()` so that we early-reject any non-instantiable
class/struct/union as a base.  Add a one-line entry to
clang/docs/ReleaseNotes.rst.  Fixes #147213.

shashforge committed:
Merge branch 'main' into fix-invalid-base-ice

shashi1687 commented:
> Thanks for working on this
> 
> Did you try to use `!isUsable` instead of IsInvalid in `ParseBaseSpecifier` ? I think that might be a simpler fix.
> 
> This change needs a release note. Please add an entry to `clang/docs/ReleaseNotes.rst` in the section the most adapted to the change, and referencing any Github issue this change fixes. Thanks!

Thank you so much for the suggestion and for taking the time to review this! 🙏  I’ve switched over to using `!BaseTy->isUsableType()` in `ParseBaseSpecifier`, dropped the old `isInvalid` checks, and added the following one-line entry under “Bug Fixes” in **clang/docs/ReleaseNotes.rst**:

```rst
* Fixed a crash when parsing an unusable type as a base specifier. (#147213)
```

I’ve also updated the commit message to:

```
ParseDeclCXX: reject unusable base classes via isUsableType

In Parser::ParseBaseSpecifier, use !BaseTy->isUsableType() to early-reject
any class/struct/union that isn’t instantiable as a base. Add a release note.
Fixes #147213.
```

Could you please take another look and let me know if anything else needs tweaking? Thanks again for your guidance!

cor3ntin commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -923,6 +923,7 @@ Bug Fixes to C++ Support
 - Improved handling of variables with ``consteval`` constructors, to
   consistently treat the initializer as manifestly constant-evaluated.
   (#GH135281)
+- Switch `ParseBaseClause` to use `BaseResult::isUsable()` instead of `isInvalid()`, fixing dropped or mis-parsed base specifiers in C++ classes. (#GH147186)
```
```suggestion
- Fix a crash in the presence of invalid base classes. (#GH147186)
```

Release notes are meant to be user-facing

cor3ntin reviewed (approved):
LGTM

shashforge committed:
Apply suggestions from code review

Adding review comment

Co-authored-by: Corentin Jabot <corentinjabot@gmail.com>

shashforge committed:
Merge branch 'main' into fix-invalid-base-ice

shashforge committed:
Merge branch 'main' into fix-invalid-base-ice

shafik commented on clang/lib/Parse/ParseDeclCXX.cpp:
```diff
@@ -2252,7 +2252,7 @@ void Parser::ParseBaseClause(Decl *ClassDecl) {
     // Parse a base-specifier.
     BaseResult Result = ParseBaseSpecifier(ClassDecl);
-    if (Result.isInvalid()) {
+    if (!Result.isUsable()) {
```
I am a bit surprised we still don't want the `BaseInfo.push_back(Result.get())` in the else branch in the good case. Why does not having this not change behavior?
